crates/cranelift/src/obj.rs

```diff
@@ -0,0 +1,551 @@
+                        addend: r.addend.wrapping_add(symbol_offset as i64),
+                    },
+                )?;
+            }
+        }
+        Ok(())
+    }
+
+    /// This function appends a nonstandard section to the object which is only
+    /// used during `CodeMemory::allocate_for_object`.
+    ///
+    /// This custom section effectively stores a `[RUNTIME_FUNCTION; N]` into
+    /// the object file itself. This way registration of unwind info can simply
+    /// pass this slice to the OS itself and there's no need to recalculate
+    /// anything on the other end of loading a module from a precompiled object.
+    fn append_windows_unwind_info(&mut self) {
+        // Currently the binary format supported here only supports
+        // little-endian for x86_64, or at least that's all where it's tested.
+        // This may need updates for other platforms.
+        assert_eq!(self.obj.architecture(), Architecture::X86_64);
+
+        // Page-align the text section so the unwind info can reside on a
+        // separate page that doesn't need executable permissions.
+        self.obj.append_section_data(self.text_section, &[], 0x1000);
+
+        let segment = self.obj.segment_name(StandardSegment::Data).to_vec();
+        let section_id = self.obj.add_section(
+            segment,
+            b"_wasmtime_winx64_unwind".to_vec(),
+            SectionKind::ReadOnlyData,
+        );
+        let mut unwind_info = Vec::with_capacity(self.windows_unwind_info.len() * 3 * 4);
+        for info in self.windows_unwind_info.iter() {
+            unwind_info.extend_from_slice(&info.begin.to_le_bytes());
+            unwind_info.extend_from_slice(&info.end.to_le_bytes());
+            unwind_info.extend_from_slice(&info.unwind_address.to_le_bytes());
+        }
+        self.obj.append_section_data(section_id, &unwind_info, 1);
+    }
+
+    /// This function appends a nonstandard section to the object which is only
+    /// used during `CodeMemory::allocate_for_object`.
+    ///
+    /// This will generate a `.eh_frame` section, but not one that can be
+    /// naively loaded. The goal of this section is that we can create the
+    /// section once here and never again does it need to change. To describe
+    /// dynamically loaded functions though each individual FDE needs to talk
+    /// about the function's absolute address that it's referencing. Naturally
+    /// we don't actually know the function's absolute address when we're
+    /// creating an object here.
+    ///
+    /// To solve this problem the FDE address encoding mode is set to
+    /// `DW_EH_PE_pcrel`. This means that the actual effective address that the
+    /// FDE describes is a relative to the address of the FDE itself. By
+    /// leveraging this relative-ness we can assume that the relative distance
+    /// between the FDE and the function it describes is constant, which should
+    /// allow us to generate an FDE ahead-of-time here.
```

Comment:
To clarify: the alternative to this whole `DW_EH_PE_pcrel` approach would be to emit relocs for the FDEs, right?

Reply:
Indeed yeah. I'll admit that I know so little about relocations I didn't really try to do this at all and I didn't feel I needed to pursue this much because I got the `pcrel` bits working. Not having relocations is somewhat beneficial in that loading these tables is that much faster (it's just what's in-memory) and in theory amenable to the hypothetical mmap-and-go world.